README.md
<!-- README.md is generated from README.Rmd. Please edit that file -->



# pbivnorm

[![Travis-CI Build Status](https://travis-ci.org/brentonk/pbivnorm.png?branch=master)](https://travis-ci.org/brentonk/pbivnorm)
[![Coverage Status](https://coveralls.io/repos/brentonk/pbivnorm/badge.svg)](https://coveralls.io/r/brentonk/pbivnorm)

`pbivnorm` is an R package containing a vectorized function to compute the cumulative distribution function of the bivariate normal distribution.  It is based on [the `mnormt` package](http://cran.r-project.org/web/packages/mnormt/index.html) by [Adelchi Azzalini](http://azzalini.stat.unipd.it/index-en.html), which uses [Fortran code](http://www.math.wsu.edu/faculty/genz/software/fort77/mvtdstpack.f) by [Alan Genz](http://www.math.wsu.edu/faculty/genz/homepage) to compute integrals of multivariate normal densities.

A call to `pbivnorm()` produces identical output to a corresponding set of
calls to `mnormt::pmnorm()`, but at lower computational cost due to
vectorization (i.e., looping in Fortran rather than in R).


```r
library("pbivnorm")
library("mnormt")
library("MASS")
library("microbenchmark")

## Simulate data
set.seed(9497)
mu <- rep(0, 2)
Sigma <- matrix(c(1, 0.5, 0.5, 1), 2, 2)
X <- mvrnorm(100, mu = mu, Sigma = Sigma)

## Confirm equality of results
all.equal(pbivnorm(X, rho = 0.5),
          apply(X, 1, pmnorm, mean = mu, varcov = Sigma))
#> [1] TRUE

## Compare speed
microbenchmark(
    pbivnorm = pbivnorm(X, rho = 0.5),
    pmnorm = apply(X, 1, pmnorm, mean = mu, varcov = Sigma)
)
#> Unit: microseconds
#>      expr     min      lq   mean  median       uq      max neval
#>  pbivnorm  100.09  104.74  119.2  114.36   123.49   252.41   100
#>    pmnorm 8971.43 9247.10 9915.4 9446.80 10133.20 14139.80   100
```


NEWS
==================================
===== The 'pbivnorm' package =====
==================================

Maintained by Brenton Kenkel (brenton.kenkel@gmail.com).


* Enabled recycling of `x` and `y` to conform with `rho` (previously, only
  `rho` could be recycled)

* Added an argument to disallow recycling


pbivnorm 0.5-1 (2012-10-31)
---------------------------

* Added NAMESPACE file


pbivnorm 0.5-0 (2011-02-21)
---------------------------

* Initial release of package


tests/testthat.R
library(testthat)
library(pbivnorm)

test_check("pbivnorm")


tests/testthat/test-matrix.r
library("pbivnorm")
library("MASS")
context("Specifying x as a matrix")

set.seed(2425)

nobs <- 100
rho <- 0.5
X <- mvrnorm(n = nobs,
             mu = c(0, 0),
             Sigma = matrix(c(1, rho, rho, 1), 2, 2))

test_that("Results are identical whether using vectors or a matrix", {
    expect_identical(pbivnorm(x = X, rho = rho),
                     pbivnorm(x = X[, 1], y = X[, 2], rho = rho))
    expect_identical(pbivnorm(x = X, rho = -rho),
                     pbivnorm(x = X[, 1], y = X[, 2], rho = -rho))
    expect_identical(pbivnorm(x = X, rho = 0),
                     pbivnorm(x = X[, 1], y = X[, 2], rho = 0))
})



tests/testthat/test-pmnorm.r
library("pbivnorm")
library("mnormt")
library("MASS")
context("Same results as mnormt::pmnorm()")

set.seed(2529)

pbiv_via_pmnorm <- function(x, rho)
{
    apply(x, 1, pmnorm, varcov = matrix(c(1, rho, rho, 1), 2, 2))
}

nobs <- 100
rho <- 0.5
X <- mvrnorm(n = nobs,
             mu = c(0, 0),
             Sigma = matrix(c(1, rho, rho, 1), 2, 2))

test_that("Results are equal between pbivnorm() and pmnorm()", {
    expect_equal(pbivnorm(x = X, rho = rho),
                 pbiv_via_pmnorm(x = X, rho = rho))
    expect_equal(pbivnorm(x = X, rho = -rho),
                 pbiv_via_pmnorm(x = X, rho = -rho))
    expect_equal(pbivnorm(x = X, rho = 0),
                 pbiv_via_pmnorm(x = X, rho = 0))
})


R/pbivnorm.r
### R code of package 'pbivnorm', version 0.5-1 (2012-10-31).
### Author: Brenton Kenkel
###
### Based on code from 'mnormt' package by Adelchi Azzalini (University of
### Padua, Italy); used under GPL

##' Calculate probabilities from the CDF of a standard bivariate normal
##' distribution.
##'
##' This function returns values identical to those of \code{biv.nt.prob} in the
##' \pkg{mnormt} package, but is vectorized to reduce the number of Fortran
##' calls required for computation of many probabilities.
##' @title Standard bivariate normal CDF
##' @param x vector of upper integration limits for the CDF.  May also be a
##' two-column matrix, in which case \code{y} should not be used.
##' @param y vector of upper integration limits, must be same length as
##' \code{x}.
##' @param rho correlation parameter.  May be a vector of the same length as
##' \code{x}, a scalar, or a vector of any other length that can be recycled to
##' conform with \code{x}.
##' @return Numeric vector of probabilities.
##' @references
##' Genz, A.  (1992).  Numerical Computation of Multivariate Normal
##' Probabilities.  \emph{J. Computational and Graphical Statist.}, \strong{1},
##' 141--149.
##'
##' Genz, A.  (1993).  Comparison of methods for the computation of multivariate
##' normal probabilities.  \emph{Computing Science and Statistics}, \strong{25},
##' 400--405.
##'
##' Genz, A.  Fortran code for \code{MVTDSTPACK} available at
##' \url{http://www.math.wsu.edu/math/faculty/genz/software/fort77/mvtdstpack.f}
##' (as of 2011-02-21).
##' @author Fortran code by Alan Genz (see references).  R interface by Brenton
##' Kenkel (\email{brenton.kenkel@@gmail.com}), based on code from Adelchi
##' Azzalini's \pkg{mnormt} package.
##' @examples
##' x <- rnorm(10)
##' y <- rnorm(10)
##' rho <- runif(10)
##'
##' pbivnorm(x, y, rho)
##'
##' X <- cbind(x, y)
##' pbivnorm(X, rho = rho)
##'
##' ## rho can be a single value
##' rho <- runif(1)
##' pbivnorm(x, y, rho)
pbivnorm <- function(x, y, rho = 0) {
    ## allow for x to be a two-column matrix
    if (length(dim(x))) {
        if (ncol(x) != 2)
            stop("'x' must have two columns if specified as a matrix")
        if (!missing(y) && !is.null(y))
            warning("'x' was specified as a matrix, so 'y' will be ignored")
        y <- x[, 2]
        x <- x[, 1]
    }

    ## sanity checks
    if (any(abs(rho) > 1))
        stop("'rho' must be a valid correlation (-1 <= rho <= 1)")
    if (length(x) != length(y))
        stop("'x' and 'y' must have same length")

    ## turn infinite integration limits into their best finite equivalents
    x <- replace(x, x == Inf, .Machine$double.xmax)
    x <- replace(x, x == -Inf, -.Machine$double.xmax)
    y <- replace(y, y == Inf, .Machine$double.xmax)
    y <- replace(y, y == -Inf, -.Machine$double.xmax)

    ## recycle rho into correl
    correl <- numeric(length(x))
    correl[] <- rho

    ## coerce arguments to proper types to be passed to fortran
    lower <- as.double(c(0, 0))
    infin <- as.integer(c(0, 0))
    uppera <- as.double(x)
    upperb <- as.double(y)
    lt <- as.integer(length(x))
    prob <- double(lt)
    correl <- as.double(correl)

    ans <- .Fortran("PBIVNORM", prob, lower, uppera, upperb, infin, correl, lt,
                    PACKAGE="pbivnorm")[[1]]
    return(ans)
}
